libclc/x86_64-unknown-linux/libspirv/math/helpers.h

```diff
@@ -0,0 +1,82 @@
+  _CLC_OVERLOAD TYPE __##NAME##_helper(TYPE);                                  \
+  _CLC_OVERLOAD TYPE __spirv_ocl_##NAME(TYPE n) { return __##NAME##_helper(n); }
+
+#define GEN_TERNARY_BUILTIN_T(NAME, TYPE)                                      \
+  _CLC_OVERLOAD TYPE __##NAME##_helper(TYPE, TYPE, TYPE);                      \
+  _CLC_OVERLOAD TYPE __spirv_ocl_##NAME(TYPE a, TYPE b, TYPE c) {              \
+    return __##NAME##_helper(a, b, c);                                         \
+  }
+#define GEN_UNARY_BUILTIN(NAME)                                                \
+  GEN_UNARY_BUILTIN_T(NAME, float)                                             \
+  GEN_UNARY_BUILTIN_T(NAME, double)
+
+#define GEN_TERNARY_BUILTIN(NAME)                                              \
+  GEN_TERNARY_BUILTIN_T(NAME, float)                                           \
+  GEN_TERNARY_BUILTIN_T(NAME, double)
+
+#else
+
+#ifndef IS_NATIVE
+#define GETNAME(ID) __spirv_ocl_##ID
+#else
+#define GETNAME(ID) __spirv_ocl_native_##ID
+#endif
+
+// Todo: fabs is the only builtin whose vector version is not named
+// __builtin_elementwise_##NAME
+#ifndef IS_FABS
+#define GEN_UNARY_VECTOR_BUILTIN(NAME, TYPE, NUM)                              \
+  _CLC_OVERLOAD TYPE##NUM GETNAME(NAME)(TYPE##NUM n) {                         \
+    return __builtin_elementwise_##NAME(n);                                    \
+  }
+#else
+#define GEN_UNARY_VECTOR_BUILTIN(NAME, TYPE, NUM)                              \
+  _CLC_OVERLOAD TYPE##NUM GETNAME(NAME)(TYPE##NUM n) {                         \
+    return __builtin_elementwise_abs(n);                                       \
+  }
+#endif
+
+#define GEN_UNARY_VECTOR_BUILTIN_T(NAME, TYPE)                                 \
+  GEN_UNARY_VECTOR_BUILTIN(NAME, TYPE, 2)                                      \
+  GEN_UNARY_VECTOR_BUILTIN(NAME, TYPE, 3)                                      \
+  GEN_UNARY_VECTOR_BUILTIN(NAME, TYPE, 4)                                      \
+  GEN_UNARY_VECTOR_BUILTIN(NAME, TYPE, 8)                                      \
+  GEN_UNARY_VECTOR_BUILTIN(NAME, TYPE, 16)
+
+#define GEN_UNARY_BUILTIN(NAME)                                                \
+  _CLC_OVERLOAD float GETNAME(NAME)(float n) {                                 \
+    return __builtin_##NAME##f(n);                                             \
+  }                                                                            \
+  _CLC_OVERLOAD double GETNAME(NAME)(double n) { return __builtin_##NAME(n); } \
+  GEN_UNARY_VECTOR_BUILTIN_T(NAME, float)                                      \
+  GEN_UNARY_VECTOR_BUILTIN_T(NAME, double)
+
+#define GEN_TERNARY_VECTOR_BUILTIN(NAME, TYPE, NUM)                            \
+  _CLC_OVERLOAD TYPE##NUM GETNAME(NAME)(TYPE##NUM n1, TYPE##NUM n2,            \
+                                        TYPE##NUM n3) {                        \
+    return __builtin_elementwise_##NAME(n1, n2, n3);                           \
```

Comment:
Are these __builtin* specific to nativecpu?

Reply:
They are part of  `clang` see [here](https://github.com/intel/llvm/blob/8fe166fb4431645b6c7a26f4d88c0564e8d1d427/clang/include/clang/Basic/Builtins.def#L685C9-L685C34)

Reply:
That's what I was thinking. It seems a bit strange to me to be calling generic clang builtins in a target specific libclc backend. Is there a good reason to do it like this?

Maybe these clang builtins better than the ones in the libclc/generic backend? Such generic impls that would be called if you don't add a target specific impl. Or could the clang builtins used here just be added to the libclc/generic impls?

Thanks

Reply:
The clang builtins should lead to the same LLVM intrinsics used in `libclc/generic`, but the way the SPIRV builtins are implemented in `generic` doesn't play well with the x86 ABI on Linux (see https://github.com/intel/llvm/pull/10970). I think you are right and it's worth considering changing the implementation in `generic`, but I'm not 100% sure about the implications of doing so.

Reply:
OK I see, makes sense, thanks for the clarification.